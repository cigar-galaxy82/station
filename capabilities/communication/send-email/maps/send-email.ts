/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable jest/no-try-expect */
/* eslint-disable jest/no-conditional-expect */
/* eslint-disable jest/no-export */

import { MappedHTTPError, Provider } from '@superfaceai/one-sdk';

import {
  CommunicationSendEmailProfile,
  SuperfaceClient,
} from '../../../../superface/sdk';

export const sendEmailTest = (providerName: string): void => {
  describe(`communication/send-email/${providerName}`, () => {
    let client: InstanceType<typeof SuperfaceClient>;
    let profile: CommunicationSendEmailProfile;
    let provider: Provider;

    beforeEach(async () => {
      client = new SuperfaceClient();
      profile = await client.getProfile('communication/send-email');
      provider = await client.getProvider(providerName);
    });

    it('should find provider', async () => {
      expect(provider).not.toBeUndefined();
    });

    describe('SendEmail', () => {
      describe('when all inputs are correct', () => {
        it('should return messageId as result', async () => {
          const result = await profile.useCases.SendEmail.perform(
            {
              from:
                process.env[
                  `${providerName.toUpperCase()}_COMMUNICATION_SENDEMAIL_FROM`
                ],
              to: process.env.COMMUNICATION_SENDEMAIL_TO,
              subject: 'Station test',
              text: `Station test - ${providerName}`,
            },
            { provider }
          );

          expect(typeof result.unwrap().messageId).toBe('string');
        });
      });

      describe('when inputs are invalid', () => {
        it('should throw on unwrap', async () => {
          const result = await profile.useCases.SendEmail.perform(
            {
              to: 'invalidemail',
              from: 'invalidemail',
              subject: '',
              text: '',
            },
            { provider }
          );

          try {
            result.unwrap();
            throw new Error('must throw something');
          } catch (error) {
            expect(error).toBeInstanceOf(MappedHTTPError);
            expect(error.properties.title).toBe('Invalid inputs');
          }
        });
      });
    });

    describe('SendTemplatedEmail', () => {
      describe('when all inputs are correct', () => {
        it('should return messagaId as result', async () => {
          const result = await profile.useCases.SendTemplatedEmail.perform(
            {
              from:
                process.env[
                  `${providerName.toUpperCase()}_COMMUNICATION_SENDEMAIL_FROM`
                ],
              to: process.env.COMMUNICATION_SENDEMAIL_TO,
              templateId:
                process.env[
                  `${providerName.toUpperCase()}_COMMUNICATION_SENDEMAIL_TEMPLATE_ID`
                ],
              templateData: {
                from:
                  process.env[
                    `${providerName.toUpperCase()}_COMMUNICATION_SENDEMAIL_FROM`
                  ],
                to: process.env.COMMUNICATION_SENDEMAIL_TO,
              },
            },
            { provider }
          );

          expect(typeof result.unwrap().messageId).toBe('string');
        });
      });

      describe('when inputs are invalid', () => {
        it('should throw error on unwrap', async () => {
          const result = await profile.useCases.SendTemplatedEmail.perform(
            {
              to: 'invalidemail',
              from: 'invalidemail',
              templateId: 'invalid',
              templateData: null,
            },
            { provider }
          );

          try {
            result.unwrap();
            throw new Error('must throw something');
          } catch (error) {
            expect(error).toBeInstanceOf(MappedHTTPError);
            expect(error.properties.title).toBe('Invalid inputs');
          }
        });
      });
    });
  });
};
