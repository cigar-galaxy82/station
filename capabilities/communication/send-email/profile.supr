"
Send E-Mail
"

name = "communication/send-email"
version = "1.0.0"

usecase SendEmail unsafe {
  input {
    from!
    to!
    subject!
    text
    html
  }

  result {
    messageId
  }

  "
  Based on https://tools.ietf.org/html/rfc7807
  "
  error {
    title
    detail
    instance
  }
}
