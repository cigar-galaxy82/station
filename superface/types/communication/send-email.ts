import { typeHelper, TypedProfile } from '@superfaceai/one-sdk';
export interface SendEmailInput {
    from: unknown;
    to: unknown;
    subject: unknown;
    text?: unknown;
    html?: unknown;
}
export interface SendEmailResult {
    messageId?: unknown;
}
export const profile = {
    "SendEmail": typeHelper<SendEmailInput, SendEmailResult>()
};
export type CommunicationSendEmailProfile = TypedProfile<typeof profile>;
export const communicationSendEmail = {
    "communication/send-email": profile
};
