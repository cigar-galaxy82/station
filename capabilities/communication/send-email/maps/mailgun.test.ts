import { sendEmailTest } from './send-email';

sendEmailTest(
  'mailgun',
  {
    from: 'demo@demo.superface.org',
    to: 'demo@superface.ai',
  },
  {
    beforeRecordingSave: recordings => {
      recordings.forEach(recording => {
        recording.path = recording.path.replace(
          process.env.MAILGUN_DOMAIN as string,
          'example.com'
        );
      });
    },
    afterRecordingLoad: scopes => {
      scopes.forEach(scope => {
        scope.filteringPath(
          new RegExp(process.env.MAILGUN_DOMAIN as string, 'g'),
          'example.com'
        );
      });
    },
  }
);
