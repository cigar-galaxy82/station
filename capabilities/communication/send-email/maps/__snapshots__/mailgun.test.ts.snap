// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`communication/send-email/mailgun SendEmail when all inputs are correct should return messageId as result 1`] = `
Object {
  "value": Object {
    "messageId": "<20211103112237.1.4BF44DCD2AB860D7@demo.superface.org>",
  },
}
`;

exports[`communication/send-email/mailgun SendEmail when inputs are invalid should throw an exception 1`] = `
Object {
  "error": "HTTPError: Expected HTTP error
Properites: {
  \\"title\\": \\"Invalid inputs\\",
  \\"detail\\": \\"'from' parameter is not a valid address. please check documentation\\"
}
AST Path: definitions[0].statements[3].responseHandlers[1].statements[0]
Original Map Location: Line 35, column 7",
}
`;
